akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala

```diff
@@ -0,0 +1,211 @@
+      if (nextElem ne null) {
+        needWakeup = false
+        if (onEvent(nextElem)) tryProcessNext(firstAttempt = true)
+      } else {
+        needWakeup = true
+        // additional poll() to grab any elements that might missed the needWakeup
+        // and have been enqueued just after it
+        if (firstAttempt)
+          tryProcessNext(firstAttempt = false)
+      }
+    }
+
+    def isShuttingDown: Boolean = shuttingDown
+
+    // External API
+    def enqueue(ev: Event): Unit = {
+      queue.add(ev)
+      if (needWakeup) {
+        needWakeup = false
+        wakeupCallback.invoke(NotUsed)
+      }
+    }
+
+    override def postStop(): Unit = {
+      // First announce that we are shutting down. This will notify late-comers to not even put anything in the queue
+      shuttingDown = true
+      // Anybody that missed the announcement needs to be notified.
+      var event = queue.poll()
+      while (event ne null) {
+        event match {
+          case Register(_, demandCallback) ⇒ demandCallback.invoke(MergeHub.Cancel)
+          case _                           ⇒
+        }
+        event = queue.poll()
+      }
+
+      // Kill everyone else
+      val states = demands.values().iterator()
+      while (states.hasNext) {
+        states.next().close()
+      }
+    }
+  }
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Sink[T, NotUsed]) = {
+    // TODO: Move inside the merge logic
+    val idCounter = new AtomicLong()
+    val producerCount = new AtomicLong()
+
+    val logic: MergedSourceLogic = new MergedSourceLogic(shape, idCounter)
+
+    val sink = new GraphStage[SinkShape[T]] {
+      val in: Inlet[T] = Inlet("MergeHub.in")
+      override val shape: SinkShape[T] = SinkShape(in)
+
+      override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) with InHandler {
+        // Start from non-zero demand to avoid initial delays.
+        // The HUB will expect this behavior.
+        private var demand: Long = perProducerBufferSize
+        private val id = idCounter.getAndIncrement()
+
+        override def preStart(): Unit = {
+          if (!logic.isShuttingDown) {
+            logic.enqueue(Register(id, getAsyncCallback(onDemand)))
+
+            // At this point, we could be in the unfortunate situation that:
+            // - we missed the shutdown announcement and entered this arm of the if statement
+            // - *before* we enqueued our Register event, the Hub already finished looking at the queue
+            //   and is now dead, so we are never notified again.
+            // To safeguard against this, we MUST check the announcement again. This is enough:
+            // if the Hub is no longer looking at the queue, then it must be that isShuttingDown must be already true.
+            if (!logic.isShuttingDown) pullWithDemand()
+            else completeStage()
```

Comment:
here we might leak a Register message in the queue. We have seen the shuttingDown flag, and we already enqueued the Register. Unfortunately, we don't know if the other side is emptying the queue right now, or already finished it. We cannot clear it if it is still draining it, so this means that we have to leave it as it is. This can mean that the queue is not guaranteed to be empty after the HUB shuts down.

Maybe one solution would be that when the Source goes away, it schedules a Runnable to the scheduler to clean up the queue after a while.